### PR TITLE
[Ansor][AutoTVM v2.0] Phase 1: Add pragma/storage_align/rfactor steps

### PR DESCRIPTION
**Pull request by jcf94**
For the full upstream plan, see [Ansor RFC](https://discuss.tvm.ai/t/rfc-ansor-an-auto-scheduler-for-tvm-autotvm-v2-0/7005/32).

In this PR, we bring pragma/storage_align/rfactor steps for Ansor auto_scheduler.

The Pragma step and StorageAlign step are two simple steps that add attributes to schedule.
The Rfactor step adds extra stage like CacheRead/CacheWrite do.

cc @merrymercy @comaniac @junrushao1994 @FrozenGene @jroesch

### CODE REVIEW

File: src/auto_scheduler/transform_step.cc [hunk 1]
```diff
@@ -967,11 +1162,27 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
...
+        }
+      }
+      // add SplitStepNode required by rfactor
+      CHECK_GE(i, 1);
+      CHECK(transform_steps[i - 1]->IsInstance<SplitStepNode>());
+      const Step& split_step = transform_steps[i - 1];
+      CHECK_EQ(split_step->stage_id, step->stage_id);
+      ret_steps.push_back(split_step);
+      // add RfactorStepNode
+      ret_steps.push_back(step);
```
Comment (jcf94) [inline src/auto_scheduler/transform_step.cc:1184]: These checks here are specialized for the Rfactor in our search policy, will the rules be kept like this?

File: include/tvm/auto_scheduler/transform_step.h [hunk 2]
```diff
@@ -812,6 +937,74 @@ class CacheWriteStep : public Step {
...
+
+/*!
+ * \brief Managed reference to RfactorStepNode.
+ * \sa RfactorStepNode
+ */
+class RfactorStep : public Step {
+ public:
+  /*!
+   * \brief The constructor.
+   * \param stage_id The index of the iterator to be factored.
```
Comment (merrymercy) [inline include/tvm/auto_scheduler/transform_step.h:992]:
```suggestion
   * \param stage_id The index of the stage to be factored.
```

File: src/auto_scheduler/transform_step.cc (hunk 1 above)
Comment (merrymercy) [reply src/auto_scheduler/transform_step.cc:1184]: yes. Otherwise, there will be errors in in creating the new dag

**Comment by jcf94** [comment]
> LGTM. Please note that #6142 changes the print state APIs, so we need to rebase one PR after merging another.

Ok, let me do the rebase after #6142.

File: tests/python/unittest/test_auto_scheduler_measure.py [hunk 3]
```diff
@@ -71,6 +73,13 @@ def test_record():
     s.compute_at(D_global, E, s[E].iters[2])
     # Cache Write
     s.cache_write(D, "shared")
+    # Pragma
```
Comment (merrymercy) [inline tests/python/unittest/test_auto_scheduler_measure.py:76]: This test is too long. Split it into smaller functions

**Comment by tqchen** [comment]
cc @junrushao1994 @jwfromm

File: tests/python/unittest/test_auto_scheduler_measure.py (hunk 3 above)
Comment (jcf94) [reply tests/python/unittest/test_auto_scheduler_measure.py]: Split to 4 functions.

**Comment by jcf94** [comment]
Thanks to all! It's really great to work with you.
We've successfully merged almost all of the Ansor basic components, except the feature extraction and cost model parts.

Next, @merrymercy will continue to add the cost model while at the same time I'll start to bring our Sketch Search Policy(proposed in our paper) to TVM, and tutorials about how to tune a op/subgraph will together to be added. Hope the PR of Sketch Search Policy will come soon.

The API design and our custom sketch rule support (which works the same as the current AutoTVM template) are ongoing to be refined. We would be happy if anyone has suggestions on this project, and you're welcomed to join us to improve Ansor.

cc @minminsun @comaniac @FrozenGene @yangjunpro @tqchen @junrushao1994 @jroesch

**Pull request merged**